contracts/pangolin-periphery/PangolinBridgeMigrationRouter.sol

```diff
@@ -0,0 +1,358 @@
+    }
+
+    /**
+     * @notice Given an address, returns whether or not he's already an admin
+     * @param account The address of the account.
+     * @return Whether or not the account address is an admin.
+     */
+    function isAdmin(address account) external view returns(bool) {
+        return adminRole.has(account);
+    }
+
+    /**
+     * @notice Given an token, and it's migrator address, it configures the migrator for later usage
+     * @param token The ERC20 token address that will be migrated using the migrator
+     * @param migrator The WrappedERC20 token address that will be migrate the token
+     */
+    function addMigrator(address token, address migrator) external onlyAdmin {
+        uint256 amount = IBridgeToken(migrator).swapSupply(token);
+        require(
+            amount > 0,
+            "The migrator doesn't have swap supply for this token"
+        );
+        _allowToken(token, migrator);
+        bridgeMigrator[token] = migrator;
+    }
+
+    /**
+     * @notice Internal function to manage approval, allows an ERC20 to be spent to the maximum
+     * @param tokenAddress The ERC20 token address that will be allowed to be used
+     * @param spenderAddress Who's going to spend the ERC20 token
+     */
+    function _allowToken(address tokenAddress, address spenderAddress) internal {
+        IPangolinERC20(tokenAddress).approve(spenderAddress, uint(-1));
+    }
+
+    /**
+     * @notice Internal function add liquidity on a low level, without the use of a router
+     * @dev This function will try to maximize one of the tokens amount and match the other
+     * one, can cause dust so consider charge backs
+     * @param pairToken The pair that will receive the liquidity
+     * @param token0 The first token of the pair
+     * @param token1 The second token of the pair
+     * @param amountIn0 The amount of first token that can be used to deposit liquidity
+     * @param amountIn1 The amount of second token that can be used to deposit liquidity
+     * @param to The address who will receive the liquidity
+     * @return amount0 Charge back from token0
+     * @return amount1 Charge back from token1
+     * @return liquidityAmount Total liquidity token amount acquired
+     */
+    function _addLiquidity(
+        address pairToken,
+        address token0,
+        address token1,
+        uint amountIn0,
+        uint amountIn1,
+        address to
+    ) private returns (uint amount0, uint amount1, uint liquidityAmount) {
+        (uint112 reserve0, uint112 reserve1,) = IPangolinPair(pairToken).getReserves();
+        uint quote0 = amountIn0;
+        uint quote1 = PangolinLibrary.quote(amountIn0, reserve0, reserve1);
+        if (quote1 > amountIn1) {
+            quote1 = amountIn1;
+            quote0 = PangolinLibrary.quote(amountIn1, reserve1, reserve0);
+        }
+        TransferHelper.safeTransfer(token0, pairToken, quote0);
+        TransferHelper.safeTransfer(token1, pairToken, quote1);
```

Comment:
Please make sure this isn't vulnerable to re-entrancy attacks

Reply:
The onlyt way I see re-entrance being executed here is if pairToken either calls addLiquidity again on getReserves or transfer method, but none of them would yield any bad outcomes, at least not one that I can foresee. This contract only works on liquidity that the user provided and does not hold any balance, we make sure we refund all balances back at the end of the operation...

Are there any specific concerns you would like to raise for the re-entrancy?